core/src/main/java/cucumber/runtime/Runtime.java

```
@@ -1,66 +1,251 @@
-                   RunnerSupplier runnerSupplier,
-                   FeatureSupplier featureSupplier
-    ) {
+    public Runtime(final Plugins plugins,
+                   final RuntimeOptions runtimeOptions,
+                   final EventBus bus,
+                   final Filters filters,
+                   final RunnerSupplier runnerSupplier,
+                   final FeatureSupplier featureSupplier,
+                   final ExecutorService executor) {
 
         this.plugins = plugins;
+        this.runtimeOptions = runtimeOptions;
         this.filters = filters;
         this.bus = bus;
-        this.runner = runnerSupplier.get();
+        this.runnerSupplier = runnerSupplier;
         this.featureSupplier = featureSupplier;
+        this.executor = executor;
+        this.exitStatus = new ExitStatus(runtimeOptions);
+        exitStatus.setEventPublisher(bus);
     }
 
     public void run() {
-        List<CucumberFeature> features = featureSupplier.get();
+        final List<CucumberFeature> features = featureSupplier.get();
         bus.send(new TestRunStarted(bus.getTime()));
         for (CucumberFeature feature : features) {
             feature.sendTestSourceRead(bus);
         }
 
-        StepDefinitionReporter stepDefinitionReporter = plugins.stepDefinitionReporter();
+        final StepDefinitionReporter stepDefinitionReporter = plugins.stepDefinitionReporter();
+        runnerSupplier.get().reportStepDefinitions(stepDefinitionReporter);
 
-        runner.reportStepDefinitions(stepDefinitionReporter);
-
-        for (CucumberFeature cucumberFeature : features) {
-            runFeature(cucumberFeature);
+        final FeatureCompiler compiler = new FeatureCompiler();
+        for (CucumberFeature feature : features) {
+            for (final PickleEvent pickleEvent : compiler.compileFeature(feature)) {
+                if (filters.matchesFilters(pickleEvent)) {
+                    executor.execute(new Runnable() {
```

If I see this correctly, that will simply swallow any exceptions thrown while executing a pickle - i.e it will log them, dispose the executor thread and then execute the next pickle in a new executor thread.
I have seen this behaviour because there's a bug in my ConcurrentEventListener. It doesn't make the cucumber run fail!

---

Can you make an issue for that? Ideally we catch all exceptions in the runner and fail the pickle.

---

See https://github.com/cucumber/cucumber-jvm/issues/1628